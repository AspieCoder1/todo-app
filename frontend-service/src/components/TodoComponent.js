import React from 'react';

import styles from '../styles/components/todoComponent.module.scss';

export default class TodoComponent extends React.Component {
	state = {
		completed: false,
	};

	componentDidMount() {
		this.setState({
			completed: this.props.todo.completed,
		});
	}

	render() {
		const { todo, removeTodo, onToggle } = this.props;
		return (
			<div className={styles.todo}>
				<input
					type="checkbox"
					name="completed"
					id="completed"
					checked={this.state.completed}
					value={this.state.completed}
					onChange={e => {
						this.setState({
							completed: e.target.checked,
						});

						onToggle(todo);
					}}
				/>
				<h3 className={styles.body}>{todo.body}</h3>
				<button className={styles.button} onClick={() => removeTodo(todo._id)}>
					Delete Todo
				</button>
			</div>
		);
	}
}
