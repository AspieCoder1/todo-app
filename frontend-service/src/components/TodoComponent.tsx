import * as React from 'react';

import styles from '../styles/components/todoComponent.module.scss';
import { ITodo } from '../../../backend-service/src/models/todo';
import { AxiosPromise, AxiosRequestConfig } from 'axios';

interface IProps {
	todo: ITodo;
	removeTodo: (
		todo: ITodo,
	) => (dispatch: any) => AxiosPromise<AxiosRequestConfig>;
	onToggle: (
		todo: ITodo,
	) => (dispatch: any) => AxiosPromise<AxiosRequestConfig>;
}

interface IState {
	completed: boolean;
}

export default class TodoComponent extends React.Component<IProps, IState> {
	state = {
		completed: false,
	};

	componentDidMount() {
		this.setState({
			completed: this.props.todo.completed,
		});
	}

	onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		const { todo, onToggle } = this.props;
		this.setState({ completed: e.target.checked });
		onToggle(todo);
	};

	render() {
		const { todo, removeTodo, onToggle } = this.props;
		return (
			<div className={styles.todo}>
				<input
					type="checkbox"
					name="completed"
					id="completed"
					checked={this.state.completed}
					// value={this.state.completed.toString()}
					onChange={this.onChange}
				/>
				<h3 className={styles.body}>{todo.body}</h3>
				<button className={styles.button} onClick={() => removeTodo(todo._id)}>
					Delete Todo
				</button>
			</div>
		);
	}
}
