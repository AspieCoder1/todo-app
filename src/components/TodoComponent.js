import React from 'react';

export default class TodoComponent extends React.Component {
	state = {
		completed: false,
	};

	componentDidMount() {
		this.setState({
			completed: this.props.todo.completed,
		});
	}

	render() {
		const { todo, removeTodo, onToggle } = this.props;
		return (
			<div>
				<input
					type="checkbox"
					name="completed"
					id="completed"
					checked={this.state.completed}
					value={this.state.completed}
					onChange={e => {
						this.setState({
							completed: e.target.checked,
						});

						onToggle(todo, this.state.completed);
					}}
				/>
				<p>COmpleted State: {`${this.state.completed}`}</p>
				<h3>{todo.body}</h3>
				<button onClick={() => removeTodo(todo._id)}>Delete Todo</button>
			</div>
		);
	}
}
