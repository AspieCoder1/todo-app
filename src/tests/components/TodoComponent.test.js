import React from 'react';
import { shallow } from 'enzyme';
import TodoComponent from '../../components/TodoComponent';

it('should render TodoComponent correctly', () => {
	const todo = {
		body: 'Test',
		completed: false,
	};
	const wrapper = shallow(
		<TodoComponent
			todo={todo}
			removeTodo={() => undefined}
			onToggle={() => undefined}
		/>
	);

	expect(wrapper).toMatchSnapshot();
});

it('should render a todo correctly', () => {
	const todo = {
		_id: 0,
		body: 'Test',
		completed: false,
	};
	const wrapper = shallow(
		<TodoComponent
			key={todo._id}
			todo={todo}
			removeTodo={() => undefined}
			onToggle={() => undefined}
		/>
	);

	expect(wrapper.find('h3').text()).toBe(todo.body);
	expect(wrapper.find('input').checked).toBeFalsy();
});

it('should update state when checkbox is ticked', () => {
	const todo = {
		body: 'Test',
		completed: false,
	};
	const wrapper = shallow(
		<TodoComponent
			todo={todo}
			removeTodo={() => undefined}
			onToggle={() => undefined}
		/>
	);
	wrapper.find('#completed').simulate('change', { target: { checked: true } });

	expect(wrapper.state().completed).toBeTruthy();
});

it('should trigger onToggle event', () => {
	const onToggle = jest.fn();

	const todo = {
		body: 'Test',
		completed: false,
	};
	const wrapper = shallow(
		<TodoComponent
			todo={todo}
			removeTodo={() => undefined}
			onToggle={onToggle}
		/>
	);
	wrapper.find('#completed').simulate('change', { target: { checked: true } });

	expect(onToggle).toHaveBeenCalledTimes(1);
	expect(onToggle).toHaveBeenCalledWith(todo, true);
});

it('should trigger removeTodo event', () => {
	const removeTodo = jest.fn();

	const todo = {
		body: 'Test',
		completed: false,
		_id: 'TEST_ID',
	};
	const wrapper = shallow(
		<TodoComponent
			todo={todo}
			removeTodo={removeTodo}
			onToggle={() => undefined}
		/>
	);
	wrapper.find('button').simulate('click');

	expect(removeTodo).toHaveBeenCalledTimes(1);
	expect(removeTodo).toHaveBeenCalledWith('TEST_ID');
});
