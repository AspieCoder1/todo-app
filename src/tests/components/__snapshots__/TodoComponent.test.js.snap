// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TodoComponent correctly 1`] = `
<div>
  <input
    checked={false}
    id="completed"
    name="completed"
    onChange={[Function]}
    type="checkbox"
    value={false}
  />
  <h3>
    Test
  </h3>
  <button
    onClick={[Function]}
  >
    Delete Todo
  </button>
</div>
`;
